File: Android/app/src/main/java/com/jiny/catchtherule/data/MockRankingService.kt
package com.jiny.catchtherule.data

import com.jiny.catchtherule.core.model.GameMode
import com.jiny.catchtherule.core.model.RankEntry
import kotlinx.coroutines.delay
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

/**
 * 서버 없이 동작하는 임시 랭킹. 실제 Minigame 서버 결합 전 UI/흐름 개발용.
 * 제출한 점수는 프로세스 메모리에만 유지된다.
 */
object MockRankingService : RankingService {

    private val mutex = Mutex()
    private val scores = mutableListOf(
        "규칙왕" to 42,
        "패턴마스터" to 38,
        "수열요정" to 35,
        "두뇌풀가동" to 31,
        "초보탈출" to 24,
        "느긋한고양이" to 19,
        "퍼즐러" to 16,
    )

    override suspend fun submit(score: Int, nickname: String, mode: GameMode): Int {
        delay(400) // 네트워크 흉내
        return mutex.withLock {
            val idx = scores.indexOfFirst { it.first == nickname }
            if (idx >= 0) {
                scores[idx] = nickname to maxOf(scores[idx].second, score)
            } else {
                scores.add(nickname to score)
            }
            scores.sortByDescending { it.second }
            scores.indexOfFirst { it.first == nickname } + 1
        }
    }

    override suspend fun leaderboard(mode: GameMode): List<RankEntry> {
        delay(300)
        return mutex.withLock {
            scores.sortedByDescending { it.second }
                // 샘플 랭킹은 한국 국기로 표시.
                .mapIndexed { i, (nick, sc) -> RankEntry(rank = i + 1, nickname = nick, score = sc, country = "KR") }
        }
    }
}

File: iOS/CatchTheRule/Services/MockRankingService.swift
import Foundation

/// 서버 없이 동작하는 임시 랭킹. 실제 Minigame 서버 결합 전 UI/흐름 개발용.
/// 제출한 점수는 프로세스 메모리에만 유지된다.
final class MockRankingService: RankingService {
    static let shared = MockRankingService()

    private actor Board {
        var scores: [(nickname: String, score: Int)] = [
            ("규칙왕", 42),
            ("패턴마스터", 38),
            ("수열요정", 35),
            ("두뇌풀가동", 31),
            ("초보탈출", 24),
            ("느긋한고양이", 19),
            ("퍼즐러", 16),
        ]

        func submit(nickname: String, score: Int) -> Int {
            if let i = scores.firstIndex(where: { $0.nickname == nickname }) {
                scores[i].score = max(scores[i].score, score)
            } else {
                scores.append((nickname, score))
            }
            scores.sort { $0.score > $1.score }
            return (scores.firstIndex { $0.nickname == nickname } ?? 0) + 1
        }

        func snapshot() -> [(nickname: String, score: Int)] {
            scores.sorted { $0.score > $1.score }
        }
    }

    private let board = Board()

    func submit(score: Int, nickname: String, mode: GameMode) async throws -> Int {
        try? await Task.sleep(nanoseconds: 400_000_000) // 네트워크 흉내
        return await board.submit(nickname: nickname, score: score)
    }

    func leaderboard(mode: GameMode) async throws -> [RankEntry] {
        try? await Task.sleep(nanoseconds: 300_000_000)
        let snap = await board.snapshot()
        return snap.enumerated().map { idx, item in
            // 샘플 랭킹은 한국 국기로 표시.
            RankEntry(rank: idx + 1, nickname: item.nickname, score: item.score, country: "KR")
        }
    }
}
